hogben/simulate.py

```diff
@@ -10,368 +10,270 @@
+                model, data = self.simulate(i)
+                models.append(model)
+                datasets.append(data)
+
+        return models, datasets
+
+    def refl1d_experiment(self, q_array: np.ndarray, spin_state: Optional[
+        int] = None) -> refl1d.experiment.Experiment:
+        """Creates a Refl1D experiment for a given `sample` and `q_array`.
+        Also calculates the resolution for refl1D to be the same as is defined
+        for refnx as it is not by default.
+
+        Args:
+            q_array: Q points to use in the experiment.
+            spin_state: spin state to simulate if given a magnetic sample.
+
+        Returns:
+            refl1d.experiment.Experiment: experiment for the given `sample`.
+
+        """
+        # Transform the resolution from refnx to Refl1D format.
+        refl1d_dq = self.dq / (100 * np.sqrt(8 * np.log(2)))
+
+        # Calculate the dq array and use it to define a QProbe.
+        dq_array = q_array * refl1d_dq
+        probe = refl1d.probe.QProbe(q_array, dq_array, intensity=self.scale,
+                                    background=self.bkg)
+        probe.dq = self.dq
+
+        # Adjust probe calculation for constant dQ/Q resolution.
+        argmin, argmax = np.argmin(q_array), np.argmax(q_array)
+        probe.calc_Qo = np.linspace(q_array[argmin] - 3.5 * dq_array[argmin],
+                                    q_array[argmax] + 3.5 * dq_array[argmax],
+                                    21 * len(q_array))
+
+        # If the sample is magnetic, create a polarised QProbe.
+        if self.sample_model.ismagnetic:
+            probes = [None] * 4
+            probes[spin_state] = probe
+            probe = refl1d.probe.PolarizedQProbe(xs=probes, name='')
+            probe.spin_state = spin_state
+
+        return refl1d.experiment.Experiment(probe=probe, sample=self.sample_model)
+
+    def reflectivity(self, q: np.ndarray) -> np.ndarray:
+        """Calculates the model reflectivity at given `q` points.
+
+        Args:
+            q: Q points to calculate reflectance at.
+
+        Returns:
+            numpy.ndarray: reflectivity for each Q point.
+
+        """
+        # If there are no data points, return an empty array.
+        if len(q) == 0:
+            return np.array([])
+
+        # Calculate the reflectance in either refnx or Refl1D.
+        if isinstance(self.sample_model, refnx.reflect.Stucture):
+            self.model = refnx.reflect.ReflectModel(self.sample_model,
+                                                    scale=self.scale,
+                                                    bkg=self.bkg,
+                                                    dq=self.dq)
+            return self.model(q)
+
+        if isinstance(self.sample_model, refl1d.model.Stack):
+            # If magnetic, use the correct spin state.
+            experiment = self.refl1d_experiment(q,
+                                                self.sample_model.probe.spin_state)
```

Comment:
Unless I'm defining my slabs/structures wrong. A `refl1d.model.Stack` structure does not have a `probe` attribute. You can probably slap it on manually, but that's probably not something you want to do if you're doing unpolarised stuff anyway.